File: app/src/main/java/com/savvasdalkitsis/gameframe/draw/view/LayersView.java
package com.savvasdalkitsis.gameframe.draw.view;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.AttributeSet;

import com.savvasdalkitsis.gameframe.draw.model.Historical;
import com.savvasdalkitsis.gameframe.draw.model.Model;

public class LayersView extends RecyclerView {

    private LayersAdapter layers;

    public LayersView(Context context) {
        super(context);
    }

    public LayersView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public LayersView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        layers = new LayersAdapter();
        setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, true));
        setHasFixedSize(true);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new MoveLayersItemHelper());
        itemTouchHelper.attachToRecyclerView(this);
        setAdapter(layers);
    }

    public void addNewLayer() {
        layers.addNewLayer();
    }

    public void bind(Historical<Model> modelHistory) {
        layers.bind(modelHistory);
    }

    private class MoveLayersItemHelper extends ItemTouchHelper.SimpleCallback {

        private final int ALLOW_DRAG = makeFlag(ItemTouchHelper.ACTION_STATE_DRAG,
                ItemTouchHelper.DOWN | ItemTouchHelper.UP | ItemTouchHelper.START | ItemTouchHelper.END);
        private boolean newMove = true;

        MoveLayersItemHelper() {
            super(ItemTouchHelper.UP | ItemTouchHelper.DOWN, 0);
        }

        @Override
        public void onMoved(RecyclerView recyclerView, ViewHolder viewHolder, int fromPos, ViewHolder target, int toPos, int x, int y) {
            super.onMoved(recyclerView, viewHolder, fromPos, target, toPos, x, y);
            layers.swapLayers(viewHolder, target);
        }

        @Override
        public boolean onMove(RecyclerView recyclerView, ViewHolder viewHolder, ViewHolder target) {
            if (newMove) {
                newMove = false;
                layers.swapStarted();
            }
            return target.getAdapterPosition() > 0;
        }

        @Override
        public void onSwiped(ViewHolder viewHolder, int direction) {
        }

        @Override
        public int getMovementFlags(RecyclerView recyclerView, ViewHolder viewHolder) {
            return viewHolder.getAdapterPosition() > 0 ? ALLOW_DRAG : 0;
        }

        @Override
        public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
            super.clearView(recyclerView, viewHolder);
            motionFinished();
        }

        private void motionFinished() {
            newMove = true;
            layers.swapLayersFinished();
        }
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/workspace/element/layer/view/LayersView.kt
package com.savvasdalkitsis.gameframe.feature.workspace.element.layer.view

import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.util.AttributeSet

import com.savvasdalkitsis.gameframe.feature.history.usecase.HistoryUseCase
import com.savvasdalkitsis.gameframe.feature.workspace.model.WorkspaceModel

class LayersView : RecyclerView {

    private lateinit var layers: LayersAdapter

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)

    override fun onFinishInflate() {
        super.onFinishInflate()
        layers = LayersAdapter()
        layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, true)
        setHasFixedSize(true)
        val itemTouchHelper = ItemTouchHelper(MoveLayersItemHelper())
        itemTouchHelper.attachToRecyclerView(this)
        adapter = layers
    }

    fun addNewLayer() = layers.addNewLayer()

    fun bind(modelHistory: HistoryUseCase<WorkspaceModel>) = layers.bind(modelHistory)

    private inner class MoveLayersItemHelper internal constructor() : ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, 0) {

        private val ALLOW_DRAG = ItemTouchHelper.Callback.makeFlag(ItemTouchHelper.ACTION_STATE_DRAG,
                ItemTouchHelper.DOWN or ItemTouchHelper.UP or ItemTouchHelper.START or ItemTouchHelper.END)
        private var newMove = true

        override fun onMoved(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, fromPos: Int, target: RecyclerView.ViewHolder, toPos: Int, x: Int, y: Int) {
            super.onMoved(recyclerView, viewHolder, fromPos, target, toPos, x, y)
            layers.swapLayers(viewHolder, target)
        }

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            if (newMove) {
                newMove = false
                layers.swapStarted()
            }
            return target.adapterPosition > 0
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}

        override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) =
                if (viewHolder.adapterPosition > 0) ALLOW_DRAG else 0

        override fun clearView(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder) {
            super.clearView(recyclerView, viewHolder)
            motionFinished()
        }

        private fun motionFinished() {
            newMove = true
            layers.swapLayersFinished()
        }
    }
}